perma_web/api2/serializers.py

```diff
@@ -0,0 +1,242 @@
+                unique_query = Folder.objects.filter(parent_id=parent_id, name=data['name'])
+                if self.instance:
+                    unique_query = unique_query.exclude(pk=self.instance.pk)
+                if unique_query.exists():
+                    raise serializers.ValidationError({'name':"A folder with that name already exists at that location."})
+        return data
+
+
+### ORGANIZATION ###
+
+class OrganizationSerializer(BaseSerializer):
+    registrar = serializers.StringRelatedField()
+    shared_folder = FolderSerializer()
+
+    class Meta:
+        model = Organization
+        fields = ('id', 'name', 'registrar', 'default_to_private', 'shared_folder')
+
+
+### CAPTUREJOB ###
+
+class CaptureJobSerializer(BaseSerializer):
+    guid = serializers.PrimaryKeyRelatedField(source='link', read_only=True)
+    class Meta:
+        model = CaptureJob
+        fields = ('guid', 'status', 'attempt', 'step_count', 'step_description', 'capture_start_time', 'capture_end_time', 'queue_position')
+
+
+### CAPTURE ###
+
+class CaptureSerializer(BaseSerializer):
+    class Meta:
+        model = Capture
+        fields = ('role', 'status', 'url', 'record_type', 'content_type', 'user_upload', 'playback_url')
+
+
+### LINK ###
+
+class LinkSerializer(BaseSerializer):
+    url = serializers.CharField(source='submitted_url', max_length=2100, required=False, allow_blank=True)
+    title = serializers.CharField(source='submitted_title', max_length=2100, required=False)
+    description = serializers.CharField(source='submitted_description', allow_blank=True, allow_null=True, max_length=300, required=False)
+
+    captures = CaptureSerializer(many=True, read_only=True)
+    queue_time = serializers.SerializerMethodField()
+    capture_time = serializers.SerializerMethodField()
+
+    class Meta:
+        model = Link
+        fields = ('guid', 'creation_timestamp', 'url', 'title', 'description', 'warc_size', 'captures', 'queue_time', 'capture_time')
+
+    def get_queue_time(self, link):
+        try:
+            delta = link.capture_job.capture_start_time - link.creation_timestamp
+            return delta.seconds
+        except:
+            return None
+
+    def get_capture_time(self, link):
+        try:
+            delta = link.capture_job.capture_end_time - link.capture_job.capture_start_time
+            return delta.seconds
+        except:
+            return None
+
+
+class AuthenticatedLinkSerializer(LinkSerializer):
+    """
+        Subclass of LinkSerializer with extra fields for users who own the link.
+    """
+    created_by = LinkUserSerializer(read_only=True)
+    organization = OrganizationSerializer(read_only=True)
+
+    class Meta(LinkSerializer.Meta):
+        fields = LinkSerializer.Meta.fields + ('notes', 'created_by', 'is_private', 'private_reason', 'archive_timestamp', 'organization')
+        allowed_update_fields = ['submitted_title', 'submitted_description', 'notes', 'is_private', 'private_reason']
```

Comment:
I presume this code already works (and I'll double check shortly), but commenting so I don't forget: in LinkSerializer, the fields are called plain "title" and "description" rather than "submitted_title" and "submitted_description" like they are in the model. I would have expected we'd need to list the Serializer field name in this whitelist, rather than the model field name.